{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3af041af-96d9-4c0c-b6ac-6a06acc2671f",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressMessages(library(ArchR))\n",
    "suppressMessages(library(Seurat))\n",
    "suppressMessages(library(Signac))\n",
    "suppressMessages(library(dplyr))\n",
    "suppressMessages(library(cowplot))\n",
    "suppressMessages(library(Nebulosa))\n",
    "suppressMessages(library(ggpubr))\n",
    "suppressMessages(library(BSgenome.Hsapiens.UCSC.hg38))\n",
    "suppressMessages(library(EnsDb.Hsapiens.v86))\n",
    "suppressMessages(library(Pando))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bc406efc-c810-489b-8f33-f9e16fbf1d6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/data/scATA/SingleCellOpenChromatin/local/VisiumHeartRevision/IntegrativeAnalysis/Lymphoid/code'"
      ],
      "text/latex": [
       "'/data/scATA/SingleCellOpenChromatin/local/VisiumHeartRevision/IntegrativeAnalysis/Lymphoid/code'"
      ],
      "text/markdown": [
       "'/data/scATA/SingleCellOpenChromatin/local/VisiumHeartRevision/IntegrativeAnalysis/Lymphoid/code'"
      ],
      "text/plain": [
       "[1] \"/data/scATA/SingleCellOpenChromatin/local/VisiumHeartRevision/IntegrativeAnalysis/Lymphoid/code\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "set.seed(42)\n",
    "getwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "54d04295-56c5-4358-b2ef-821ff5d7d0bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "An object of class Seurat \n",
       "159644 features across 3473 samples within 3 assays \n",
       "Active assay: RNA (28933 features, 2000 variable features)\n",
       " 2 other assays present: peaks, GeneActivity\n",
       " 5 dimensional reductions calculated: pca, umap, harmony, umap_harmony, umap_harmony_v2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "coembed <- readRDS(\"../data/coembed/coembed.annotation.Rds\")\n",
    "coembed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "20ec9ce5-83a1-4535-a7ad-f83cebabef23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "An object of class Seurat \n",
       "159644 features across 858 samples within 3 assays \n",
       "Active assay: RNA (28933 features, 2000 variable features)\n",
       " 2 other assays present: peaks, GeneActivity\n",
       " 5 dimensional reductions calculated: pca, umap, harmony, umap_harmony, umap_harmony_v2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "An object of class Seurat \n",
       "159644 features across 2615 samples within 3 assays \n",
       "Active assay: RNA (28933 features, 2000 variable features)\n",
       " 2 other assays present: peaks, GeneActivity\n",
       " 5 dimensional reductions calculated: pca, umap, harmony, umap_harmony, umap_harmony_v2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "obj.atac <- subset(coembed, tech == \"ATAC\")\n",
    "obj.rna <- subset(coembed, tech == \"RNA\")\n",
    "\n",
    "obj.atac\n",
    "obj.rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "12048832-9cfd-42f4-bbce-ea3ea0364920",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ATAC</th><th scope=col>RNA</th><th scope=col>cell_name</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>CK171#TCACCACAGCCATCAT-1</td><td>CAACCTCTCGGTTCAA-1_1_1_1    </td><td>cell-1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>CK171#GAGACTTGTAAGCCGA-1</td><td>TCACGCTAGCGATGCA-1_1_1_1_1  </td><td>cell-2</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>CK171#GACCCAGCATTTAGGC-1</td><td>CGTTCTGTCACTCGAA-1_2_1_1_1_1</td><td>cell-3</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>CK171#AAGGTTCGTCGCTACG-1</td><td>AGGACTTTCCCATAGA-1_1_1_1_1  </td><td>cell-4</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>CK171#AATGGCTGTTACGAAA-1</td><td>GGTTAACTCGCTCTAC-1_2_1      </td><td>cell-5</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>CK171#AAAGATGAGCCTGTAT-1</td><td>TGTCCTGTCTCCGAGG-1_2_1_1_1_1</td><td>cell-6</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & ATAC & RNA & cell\\_name\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & CK171\\#TCACCACAGCCATCAT-1 & CAACCTCTCGGTTCAA-1\\_1\\_1\\_1     & cell-1\\\\\n",
       "\t2 & CK171\\#GAGACTTGTAAGCCGA-1 & TCACGCTAGCGATGCA-1\\_1\\_1\\_1\\_1   & cell-2\\\\\n",
       "\t3 & CK171\\#GACCCAGCATTTAGGC-1 & CGTTCTGTCACTCGAA-1\\_2\\_1\\_1\\_1\\_1 & cell-3\\\\\n",
       "\t4 & CK171\\#AAGGTTCGTCGCTACG-1 & AGGACTTTCCCATAGA-1\\_1\\_1\\_1\\_1   & cell-4\\\\\n",
       "\t5 & CK171\\#AATGGCTGTTACGAAA-1 & GGTTAACTCGCTCTAC-1\\_2\\_1       & cell-5\\\\\n",
       "\t6 & CK171\\#AAAGATGAGCCTGTAT-1 & TGTCCTGTCTCCGAGG-1\\_2\\_1\\_1\\_1\\_1 & cell-6\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 3\n",
       "\n",
       "| <!--/--> | ATAC &lt;chr&gt; | RNA &lt;chr&gt; | cell_name &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | CK171#TCACCACAGCCATCAT-1 | CAACCTCTCGGTTCAA-1_1_1_1     | cell-1 |\n",
       "| 2 | CK171#GAGACTTGTAAGCCGA-1 | TCACGCTAGCGATGCA-1_1_1_1_1   | cell-2 |\n",
       "| 3 | CK171#GACCCAGCATTTAGGC-1 | CGTTCTGTCACTCGAA-1_2_1_1_1_1 | cell-3 |\n",
       "| 4 | CK171#AAGGTTCGTCGCTACG-1 | AGGACTTTCCCATAGA-1_1_1_1_1   | cell-4 |\n",
       "| 5 | CK171#AATGGCTGTTACGAAA-1 | GGTTAACTCGCTCTAC-1_2_1       | cell-5 |\n",
       "| 6 | CK171#AAAGATGAGCCTGTAT-1 | TGTCCTGTCTCCGAGG-1_2_1_1_1_1 | cell-6 |\n",
       "\n"
      ],
      "text/plain": [
       "  ATAC                     RNA                          cell_name\n",
       "1 CK171#TCACCACAGCCATCAT-1 CAACCTCTCGGTTCAA-1_1_1_1     cell-1   \n",
       "2 CK171#GAGACTTGTAAGCCGA-1 TCACGCTAGCGATGCA-1_1_1_1_1   cell-2   \n",
       "3 CK171#GACCCAGCATTTAGGC-1 CGTTCTGTCACTCGAA-1_2_1_1_1_1 cell-3   \n",
       "4 CK171#AAGGTTCGTCGCTACG-1 AGGACTTTCCCATAGA-1_1_1_1_1   cell-4   \n",
       "5 CK171#AATGGCTGTTACGAAA-1 GGTTAACTCGCTCTAC-1_2_1       cell-5   \n",
       "6 CK171#AAAGATGAGCCTGTAT-1 TGTCCTGTCTCCGAGG-1_2_1_1_1_1 cell-6   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_cell_pairing <- readRDS(\"../data/coembed/ATAC_RNA_matching.rds\")\n",
    "df_cell_pairing$cell_name <- paste0(\"cell-\", 1:nrow(df_cell_pairing))\n",
    "\n",
    "head(df_cell_pairing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9346a845-31d8-4c03-b681-5e6963c96e81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "An object of class Seurat \n",
       "140618 features across 858 samples within 2 assays \n",
       "Active assay: peaks (111685 features, 0 variable features)\n",
       " 1 other assay present: RNA"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "atac_counts <- obj.atac@assays$peaks@counts[, df_cell_pairing$ATAC]\n",
    "rna_counts <- obj.rna@assays$RNA@counts[, df_cell_pairing$RNA]\n",
    "\n",
    "## unify the cell names in ATAC and RNA\n",
    "colnames(atac_counts) <- df_cell_pairing$cell_name\n",
    "colnames(rna_counts) <- df_cell_pairing$cell_name\n",
    "\n",
    "## create seurat object for Pando\n",
    "obj <- CreateSeuratObject(counts = rna_counts, assay = \"RNA\")\n",
    "\n",
    "obj[['peaks']] <- CreateChromatinAssay(counts = atac_counts,\n",
    "                                      genome = \"hg38\")\n",
    "\n",
    "DefaultAssay(obj) <- \"peaks\"\n",
    "\n",
    "obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9d42fb0c-ef0e-490e-ab5d-1a32ae5b8335",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n"
     ]
    }
   ],
   "source": [
    "## add annotation for each peak\n",
    "# extract gene annotations from EnsDb\n",
    "annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86, verbose = FALSE)\n",
    "seqlevelsStyle(annotations) <- 'UCSC'\n",
    "genome(annotations) <- \"hg38\"\n",
    "\n",
    "# add the gene information to the object\n",
    "Annotation(obj[[\"peaks\"]]) <- annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6e994557-307d-4614-94dc-4212400a6436",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select variable features\n",
    "DefaultAssay(obj) <- \"RNA\"\n",
    "\n",
    "obj <- obj %>%\n",
    "    NormalizeData() %>%\n",
    "    FindVariableFeatures()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b7f00bfe-5209-497d-bebf-9f078b348c22",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initiate GRN object and select candidate regions\n",
    "obj <- initiate_grn(obj, exclude_exons = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "734943c0-a7bc-439e-b9a4-f6f444b1370d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>motif</th><th scope=col>tf</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>MA0030.1</th><td>MA0030.1</td><td>FOXF2      </td></tr>\n",
       "\t<tr><th scope=row>MA0031.1</th><td>MA0031.1</td><td>FOXD1      </td></tr>\n",
       "\t<tr><th scope=row>MA0051.1</th><td>MA0051.1</td><td>IRF2       </td></tr>\n",
       "\t<tr><th scope=row>MA0057.1</th><td>MA0057.1</td><td>MZF1(var.2)</td></tr>\n",
       "\t<tr><th scope=row>MA0059.1</th><td>MA0059.1</td><td>MAX::MYC   </td></tr>\n",
       "\t<tr><th scope=row>MA0066.1</th><td>MA0066.1</td><td>PPARG      </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & motif & tf\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\tMA0030.1 & MA0030.1 & FOXF2      \\\\\n",
       "\tMA0031.1 & MA0031.1 & FOXD1      \\\\\n",
       "\tMA0051.1 & MA0051.1 & IRF2       \\\\\n",
       "\tMA0057.1 & MA0057.1 & MZF1(var.2)\\\\\n",
       "\tMA0059.1 & MA0059.1 & MAX::MYC   \\\\\n",
       "\tMA0066.1 & MA0066.1 & PPARG      \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | motif &lt;chr&gt; | tf &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| MA0030.1 | MA0030.1 | FOXF2       |\n",
       "| MA0031.1 | MA0031.1 | FOXD1       |\n",
       "| MA0051.1 | MA0051.1 | IRF2        |\n",
       "| MA0057.1 | MA0057.1 | MZF1(var.2) |\n",
       "| MA0059.1 | MA0059.1 | MAX::MYC    |\n",
       "| MA0066.1 | MA0066.1 | PPARG       |\n",
       "\n"
      ],
      "text/plain": [
       "         motif    tf         \n",
       "MA0030.1 MA0030.1 FOXF2      \n",
       "MA0031.1 MA0031.1 FOXD1      \n",
       "MA0051.1 MA0051.1 IRF2       \n",
       "MA0057.1 MA0057.1 MZF1(var.2)\n",
       "MA0059.1 MA0059.1 MAX::MYC   \n",
       "MA0066.1 MA0066.1 PPARG      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "633"
      ],
      "text/latex": [
       "633"
      ],
      "text/markdown": [
       "633"
      ],
      "text/plain": [
       "[1] 633"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## get motifs\n",
    "suppressMessages(library(JASPAR2020))\n",
    "suppressMessages(library(TFBSTools))\n",
    "\n",
    "opts <- list()\n",
    "opts[\"species\"] <- \"Homo sapiens\"\n",
    "opts[\"collection\"] <- \"CORE\"\n",
    "motifs <- getMatrixSet(JASPAR2020, opts)\n",
    "\n",
    "motif_tfs <- as.data.frame(name(motifs))\n",
    "motif_tfs$motif <- rownames(motif_tfs)\n",
    "colnames(motif_tfs)  <- c(\"tf\", \"motif\")\n",
    "\n",
    "motif_tfs <- motif_tfs[, c(\"motif\", \"tf\")]\n",
    "\n",
    "head(motif_tfs)\n",
    "\n",
    "nrow(motif_tfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "32d6c670-93d5-4d99-acf2-fd541acf75d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Adding TF info\n",
      "\n",
      "Building motif matrix\n",
      "\n",
      "Finding motif positions\n",
      "\n",
      "Creating Motif object\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Scan candidate regions for TF binding motifs\n",
    "obj <- find_motifs(\n",
    "    obj,\n",
    "    pfm = motifs,\n",
    "    motif_tfs = motif_tfs,\n",
    "    genome = BSgenome.Hsapiens.UCSC.hg38\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "43678e35-bf72-4150-a1b6-7ecdfb242802",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Selecting candidate regulatory regions near genes\n",
      "\n",
      "Preparing model input\n",
      "\n",
      "Fitting models for 1462 target genes\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Infer gene regulatory network\n",
    "obj <- infer_grn(obj,\n",
    "                peak_to_gene_method = 'Signac',\n",
    "                 method = 'glm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b69479f7-a25b-446e-8feb-74cc0cf5d966",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>tf</th><th scope=col>target</th><th scope=col>region</th><th scope=col>term</th><th scope=col>estimate</th><th scope=col>std_err</th><th scope=col>statistic</th><th scope=col>pval</th><th scope=col>padj</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>16</th><td>MEIS3</td><td>BCOR    </td><td>chrX-40098312-40098812  </td><td>chrX_40098312_40098812:MEIS3  </td><td>0.6205188</td><td>0.09470899</td><td> 6.551847</td><td>9.801397e-11</td><td>1.108234e-09</td></tr>\n",
       "\t<tr><th scope=row>34</th><td>USF1 </td><td>MTMR8   </td><td>chrX-64395225-64395725  </td><td>chrX_64395225_64395725:USF1   </td><td>0.4564577</td><td>0.12448931</td><td> 3.666642</td><td>2.609154e-04</td><td>1.640040e-03</td></tr>\n",
       "\t<tr><th scope=row>41</th><td>BATF </td><td>MSN     </td><td>chrX-65723649-65724149  </td><td>BATF:chrX_65723649_65724149   </td><td>0.5986432</td><td>0.22571948</td><td> 2.652156</td><td>8.147016e-03</td><td>4.055076e-02</td></tr>\n",
       "\t<tr><th scope=row>59</th><td>SPI1 </td><td>PAK3    </td><td>chrX-111098751-111099251</td><td>chrX_111098751_111099251:SPI1 </td><td>0.3251020</td><td>0.06202882</td><td> 5.241145</td><td>2.011751e-07</td><td>1.752317e-06</td></tr>\n",
       "\t<tr><th scope=row>60</th><td>MGA  </td><td>PAK3    </td><td>chrX-111098751-111099251</td><td>chrX_111098751_111099251:MGA  </td><td>0.1698445</td><td>0.03296069</td><td> 5.152942</td><td>3.185931e-07</td><td>2.731812e-06</td></tr>\n",
       "\t<tr><th scope=row>66</th><td>KLF11</td><td>CXorf40A</td><td>chrX-149540298-149540798</td><td>chrX_149540298_149540798:KLF11</td><td>0.5064376</td><td>0.04789442</td><td>10.574042</td><td>1.209418e-24</td><td>2.638108e-23</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & tf & target & region & term & estimate & std\\_err & statistic & pval & padj\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t16 & MEIS3 & BCOR     & chrX-40098312-40098812   & chrX\\_40098312\\_40098812:MEIS3   & 0.6205188 & 0.09470899 &  6.551847 & 9.801397e-11 & 1.108234e-09\\\\\n",
       "\t34 & USF1  & MTMR8    & chrX-64395225-64395725   & chrX\\_64395225\\_64395725:USF1    & 0.4564577 & 0.12448931 &  3.666642 & 2.609154e-04 & 1.640040e-03\\\\\n",
       "\t41 & BATF  & MSN      & chrX-65723649-65724149   & BATF:chrX\\_65723649\\_65724149    & 0.5986432 & 0.22571948 &  2.652156 & 8.147016e-03 & 4.055076e-02\\\\\n",
       "\t59 & SPI1  & PAK3     & chrX-111098751-111099251 & chrX\\_111098751\\_111099251:SPI1  & 0.3251020 & 0.06202882 &  5.241145 & 2.011751e-07 & 1.752317e-06\\\\\n",
       "\t60 & MGA   & PAK3     & chrX-111098751-111099251 & chrX\\_111098751\\_111099251:MGA   & 0.1698445 & 0.03296069 &  5.152942 & 3.185931e-07 & 2.731812e-06\\\\\n",
       "\t66 & KLF11 & CXorf40A & chrX-149540298-149540798 & chrX\\_149540298\\_149540798:KLF11 & 0.5064376 & 0.04789442 & 10.574042 & 1.209418e-24 & 2.638108e-23\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 9\n",
       "\n",
       "| <!--/--> | tf &lt;chr&gt; | target &lt;chr&gt; | region &lt;chr&gt; | term &lt;chr&gt; | estimate &lt;dbl&gt; | std_err &lt;dbl&gt; | statistic &lt;dbl&gt; | pval &lt;dbl&gt; | padj &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| 16 | MEIS3 | BCOR     | chrX-40098312-40098812   | chrX_40098312_40098812:MEIS3   | 0.6205188 | 0.09470899 |  6.551847 | 9.801397e-11 | 1.108234e-09 |\n",
       "| 34 | USF1  | MTMR8    | chrX-64395225-64395725   | chrX_64395225_64395725:USF1    | 0.4564577 | 0.12448931 |  3.666642 | 2.609154e-04 | 1.640040e-03 |\n",
       "| 41 | BATF  | MSN      | chrX-65723649-65724149   | BATF:chrX_65723649_65724149    | 0.5986432 | 0.22571948 |  2.652156 | 8.147016e-03 | 4.055076e-02 |\n",
       "| 59 | SPI1  | PAK3     | chrX-111098751-111099251 | chrX_111098751_111099251:SPI1  | 0.3251020 | 0.06202882 |  5.241145 | 2.011751e-07 | 1.752317e-06 |\n",
       "| 60 | MGA   | PAK3     | chrX-111098751-111099251 | chrX_111098751_111099251:MGA   | 0.1698445 | 0.03296069 |  5.152942 | 3.185931e-07 | 2.731812e-06 |\n",
       "| 66 | KLF11 | CXorf40A | chrX-149540298-149540798 | chrX_149540298_149540798:KLF11 | 0.5064376 | 0.04789442 | 10.574042 | 1.209418e-24 | 2.638108e-23 |\n",
       "\n"
      ],
      "text/plain": [
       "   tf    target   region                   term                          \n",
       "16 MEIS3 BCOR     chrX-40098312-40098812   chrX_40098312_40098812:MEIS3  \n",
       "34 USF1  MTMR8    chrX-64395225-64395725   chrX_64395225_64395725:USF1   \n",
       "41 BATF  MSN      chrX-65723649-65724149   BATF:chrX_65723649_65724149   \n",
       "59 SPI1  PAK3     chrX-111098751-111099251 chrX_111098751_111099251:SPI1 \n",
       "60 MGA   PAK3     chrX-111098751-111099251 chrX_111098751_111099251:MGA  \n",
       "66 KLF11 CXorf40A chrX-149540298-149540798 chrX_149540298_149540798:KLF11\n",
       "   estimate  std_err    statistic pval         padj        \n",
       "16 0.6205188 0.09470899  6.551847 9.801397e-11 1.108234e-09\n",
       "34 0.4564577 0.12448931  3.666642 2.609154e-04 1.640040e-03\n",
       "41 0.5986432 0.22571948  2.652156 8.147016e-03 4.055076e-02\n",
       "59 0.3251020 0.06202882  5.241145 2.011751e-07 1.752317e-06\n",
       "60 0.1698445 0.03296069  5.152942 3.185931e-07 2.731812e-06\n",
       "66 0.5064376 0.04789442 10.574042 1.209418e-24 2.638108e-23"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df <- as.data.frame(obj@grn@networks$glm_network@coefs) %>%\n",
    "    subset(padj < 0.05)\n",
    "\n",
    "head(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "260667ce-7120-4468-b845-6189e6b5ac5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1007"
      ],
      "text/latex": [
       "1007"
      ],
      "text/markdown": [
       "1007"
      ],
      "text/plain": [
       "[1] 1007"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(obj@grn@networks$glm_network@coefs$target))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "cb6c918b-744b-42cf-bd04-66d37098d22b",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj <- find_modules(obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b4b047e9-4d7d-4a52-b8c8-eb22d96f3d26",
   "metadata": {},
   "outputs": [],
   "source": [
    "modules <- NetworkModules(obj)\n",
    "#modules@meta\n",
    "#modules@features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e7612df8-53a7-4eb9-945c-1642dc7d9766",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 148 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>target</th><th scope=col>tf</th><th scope=col>estimate</th><th scope=col>n_regions</th><th scope=col>n_genes</th><th scope=col>n_tfs</th><th scope=col>regions</th><th scope=col>pval</th><th scope=col>padj</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>HMBS     </td><td>CREB3</td><td>0.502456918</td><td>1</td><td>1</td><td>1</td><td>chr11-119057030-119057530                                           </td><td> 1.413769e-16</td><td> 2.239922e-15</td></tr>\n",
       "\t<tr><td>TNNT3    </td><td>E2F6 </td><td>0.150808737</td><td>2</td><td>1</td><td>2</td><td>chr11-1839764-1840264                                               </td><td> 7.670730e-35</td><td> 2.391280e-33</td></tr>\n",
       "\t<tr><td>BBC3     </td><td>EGR2 </td><td>0.221948001</td><td>3</td><td>1</td><td>3</td><td>chr19-47274053-47274553                                             </td><td> 2.391756e-05</td><td> 1.727628e-04</td></tr>\n",
       "\t<tr><td>KEAP1    </td><td>EGR3 </td><td>1.008195189</td><td>3</td><td>1</td><td>2</td><td>chr19-10478848-10479348;chr19-10602148-10602648                     </td><td> 9.042581e-10</td><td> 9.564268e-09</td></tr>\n",
       "\t<tr><td>GZMB     </td><td>EHF  </td><td>0.644146878</td><td>4</td><td>1</td><td>3</td><td>chr14-24695188-24695688                                             </td><td> 2.421297e-10</td><td> 2.663427e-09</td></tr>\n",
       "\t<tr><td>TOP1MT   </td><td>ELF3 </td><td>0.415747555</td><td>3</td><td>1</td><td>3</td><td>chr8-143281493-143281993                                            </td><td> 3.595062e-18</td><td> 6.179013e-17</td></tr>\n",
       "\t<tr><td>LINC01569</td><td>ELK3 </td><td>0.013216023</td><td>3</td><td>2</td><td>2</td><td>chr16-4307497-4307997                                               </td><td> 8.634882e-04</td><td> 5.052324e-03</td></tr>\n",
       "\t<tr><td>SPSB2    </td><td>ELK3 </td><td>0.003256462</td><td>3</td><td>2</td><td>2</td><td>chr12-6943749-6944249;chr12-6970506-6971006                         </td><td> 6.425397e-08</td><td> 5.882944e-07</td></tr>\n",
       "\t<tr><td>FCN3     </td><td>ELK4 </td><td>0.061526038</td><td>2</td><td>3</td><td>2</td><td>chr1-27321907-27322407                                              </td><td> 3.623591e-05</td><td> 2.557763e-04</td></tr>\n",
       "\t<tr><td>SNHG15   </td><td>ELK4 </td><td>0.256699658</td><td>4</td><td>3</td><td>3</td><td>chr7-44999794-45000294;chr7-45000351-45000851                       </td><td> 5.909617e-04</td><td> 3.546691e-03</td></tr>\n",
       "\t<tr><td>STAP2    </td><td>ELK4 </td><td>0.035338163</td><td>2</td><td>3</td><td>2</td><td>chr19-4400389-4400889                                               </td><td> 3.732396e-05</td><td> 2.631818e-04</td></tr>\n",
       "\t<tr><td>STOML2   </td><td>ESR1 </td><td>0.182048769</td><td>3</td><td>1</td><td>2</td><td>chr9-35080311-35080811;chr9-35095760-35096260                       </td><td> 8.314858e-34</td><td> 2.462477e-32</td></tr>\n",
       "\t<tr><td>CCR7     </td><td>ESR2 </td><td>0.429512770</td><td>4</td><td>3</td><td>3</td><td>chr17-40582251-40582751                                             </td><td> 2.734040e-11</td><td> 3.209162e-10</td></tr>\n",
       "\t<tr><td>FCGR3A   </td><td>ESR2 </td><td>0.341330263</td><td>5</td><td>3</td><td>5</td><td>chr1-161545274-161545774                                            </td><td> 4.870036e-14</td><td> 6.744474e-13</td></tr>\n",
       "\t<tr><td>NLGN2    </td><td>ESR2 </td><td>0.110788987</td><td>3</td><td>3</td><td>1</td><td>chr17-7313890-7314390;chr17-7351839-7352339;chr17-7352410-7352910   </td><td> 1.293830e-31</td><td> 3.596567e-30</td></tr>\n",
       "\t<tr><td>SYTL3    </td><td>ETS1 </td><td>0.183452614</td><td>1</td><td>1</td><td>1</td><td>chr6-158643111-158643611                                            </td><td> 3.694966e-03</td><td> 1.942064e-02</td></tr>\n",
       "\t<tr><td>PTCSC2   </td><td>ETS2 </td><td>0.498485796</td><td>1</td><td>1</td><td>1</td><td>chr9-97904628-97905128                                              </td><td> 6.605345e-51</td><td> 3.413501e-49</td></tr>\n",
       "\t<tr><td>DAXX     </td><td>ETV4 </td><td>0.491687344</td><td>8</td><td>1</td><td>3</td><td>chr6-33249651-33250151                                              </td><td> 4.476906e-14</td><td> 6.222414e-13</td></tr>\n",
       "\t<tr><td>AFMID    </td><td>ETV5 </td><td>0.212461085</td><td>3</td><td>1</td><td>3</td><td>chr17-78130482-78130982                                             </td><td> 2.535156e-05</td><td> 1.820961e-04</td></tr>\n",
       "\t<tr><td>CCR7     </td><td>ETV6 </td><td>0.130363935</td><td>4</td><td>1</td><td>3</td><td>chr17-40565166-40565666;chr17-40647845-40648345                     </td><td> 1.798356e-03</td><td> 9.912200e-03</td></tr>\n",
       "\t<tr><td>SERINC5  </td><td>FOXP1</td><td>0.314644062</td><td>3</td><td>1</td><td>2</td><td>chr5-80186588-80187088                                              </td><td> 5.482104e-04</td><td> 3.302989e-03</td></tr>\n",
       "\t<tr><td>CTU2     </td><td>FOXP2</td><td>0.064735297</td><td>8</td><td>1</td><td>2</td><td>chr16-88618912-88619412;chr16-88710799-88711299                     </td><td> 1.936689e-21</td><td> 3.784900e-20</td></tr>\n",
       "\t<tr><td>ATP6V1G2 </td><td>GABPA</td><td>0.188905241</td><td>5</td><td>1</td><td>4</td><td>chr6-31541808-31542308;chr6-31569306-31569806;chr6-31576740-31577240</td><td> 2.638782e-04</td><td> 1.657310e-03</td></tr>\n",
       "\t<tr><td>BBC3     </td><td>GFI1 </td><td>0.492520606</td><td>3</td><td>2</td><td>3</td><td>chr19-47261875-47262375                                             </td><td> 2.135358e-12</td><td> 2.722228e-11</td></tr>\n",
       "\t<tr><td>FBXO2    </td><td>GFI1 </td><td>0.358505134</td><td>1</td><td>2</td><td>1</td><td>chr1-11708563-11709063                                              </td><td> 2.324456e-30</td><td> 6.150622e-29</td></tr>\n",
       "\t<tr><td>AFMID    </td><td>GLI2 </td><td>0.404393915</td><td>3</td><td>1</td><td>3</td><td>chr17-78120864-78121364                                             </td><td> 3.210560e-34</td><td> 9.732799e-33</td></tr>\n",
       "\t<tr><td>ZNF843   </td><td>GLIS2</td><td>0.321331869</td><td>4</td><td>1</td><td>4</td><td>chr16-31458886-31459386                                             </td><td>3.321651e-191</td><td>1.967439e-188</td></tr>\n",
       "\t<tr><td>BBC3     </td><td>HIC2 </td><td>0.666970140</td><td>3</td><td>1</td><td>3</td><td>chr19-47129956-47130456                                             </td><td> 2.760745e-15</td><td> 4.088026e-14</td></tr>\n",
       "\t<tr><td>CREB3L4  </td><td>HOXB2</td><td>0.498664799</td><td>3</td><td>1</td><td>2</td><td>chr1-153945788-153946288                                            </td><td> 7.034193e-72</td><td> 6.298056e-70</td></tr>\n",
       "\t<tr><td>FCGR3A   </td><td>HOXB4</td><td>0.828054849</td><td>5</td><td>1</td><td>5</td><td>chr1-161550372-161550872                                            </td><td> 6.479635e-08</td><td> 5.918528e-07</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>MORN3   </td><td>ZBTB14 </td><td> 0.8318287039</td><td>4</td><td>1</td><td>3</td><td>chr12-121560906-121561406;chr12-121658845-121659345                                                                         </td><td> 1.133754e-13</td><td> 1.542387e-12</td></tr>\n",
       "\t<tr><td>DALRD3  </td><td>ZBTB18 </td><td> 1.0167926147</td><td>2</td><td>2</td><td>2</td><td>chr3-49021835-49022335                                                                                                      </td><td> 1.715073e-05</td><td> 1.260120e-04</td></tr>\n",
       "\t<tr><td>FCN3    </td><td>ZBTB18 </td><td> 0.6407947401</td><td>2</td><td>2</td><td>2</td><td>chr1-27343258-27343758                                                                                                      </td><td> 1.338906e-38</td><td> 4.863006e-37</td></tr>\n",
       "\t<tr><td>GSTO2   </td><td>ZBTB33 </td><td> 0.4757434981</td><td>1</td><td>1</td><td>1</td><td>chr10-104268851-104269351                                                                                                   </td><td> 7.522957e-40</td><td> 2.825696e-38</td></tr>\n",
       "\t<tr><td>DALRD3  </td><td>ZBTB6  </td><td> 0.9895490015</td><td>2</td><td>2</td><td>2</td><td>chr3-48989626-48990126                                                                                                      </td><td> 1.045595e-23</td><td> 2.205776e-22</td></tr>\n",
       "\t<tr><td>VAMP5   </td><td>ZBTB6  </td><td> 0.4881793017</td><td>2</td><td>2</td><td>2</td><td>chr2-85561906-85562406                                                                                                      </td><td> 2.940189e-11</td><td> 3.440647e-10</td></tr>\n",
       "\t<tr><td>DAXX    </td><td>ZKSCAN5</td><td> 0.9722396972</td><td>8</td><td>2</td><td>3</td><td>chr6-33271742-33272242;chr6-33279915-33280415;chr6-33298424-33298924;chr6-33322893-33323393;chr6-33425773-33426273          </td><td> 1.512623e-17</td><td> 2.488718e-16</td></tr>\n",
       "\t<tr><td>KLF16   </td><td>ZKSCAN5</td><td> 0.5000000000</td><td>4</td><td>2</td><td>2</td><td>chr19-1848488-1848988;chr19-1873393-1873893;chr19-1882818-1883318                                                           </td><td> 0.000000e+00</td><td> 0.000000e+00</td></tr>\n",
       "\t<tr><td>C16orf74</td><td>ZNF135 </td><td> 0.3696168578</td><td>2</td><td>2</td><td>2</td><td>chr16-85611569-85612069                                                                                                     </td><td> 2.278164e-27</td><td> 5.447784e-26</td></tr>\n",
       "\t<tr><td>TNNT3   </td><td>ZNF135 </td><td> 0.9969535196</td><td>2</td><td>2</td><td>2</td><td>chr11-1878260-1878760                                                                                                       </td><td> 1.632778e-29</td><td> 4.190796e-28</td></tr>\n",
       "\t<tr><td>TOP1MT  </td><td>ZNF143 </td><td> 0.5171160910</td><td>3</td><td>1</td><td>3</td><td>chr8-143246465-143246965;chr8-143290355-143290855                                                                           </td><td> 8.969845e-51</td><td> 4.574027e-49</td></tr>\n",
       "\t<tr><td>CLIC3   </td><td>ZNF148 </td><td> 0.7189429748</td><td>7</td><td>1</td><td>2</td><td>chr9-136949884-136950384;chr9-136958911-136959411;chr9-136960139-136960639;chr9-136960745-136961245;chr9-136970157-136970657</td><td> 1.672711e-10</td><td> 1.861253e-09</td></tr>\n",
       "\t<tr><td>FDXR    </td><td>ZNF16  </td><td> 0.7657769925</td><td>2</td><td>2</td><td>2</td><td>chr17-74871275-74871775                                                                                                     </td><td> 7.373991e-34</td><td> 2.192268e-32</td></tr>\n",
       "\t<tr><td>IRGQ    </td><td>ZNF16  </td><td> 1.2608986329</td><td>3</td><td>2</td><td>3</td><td>chr19-43504488-43504988                                                                                                     </td><td> 4.920331e-30</td><td> 1.279951e-28</td></tr>\n",
       "\t<tr><td>C2CD2L  </td><td>ZNF263 </td><td> 0.8435557001</td><td>5</td><td>3</td><td>3</td><td>chr11-119057030-119057530;chr11-119100173-119100673;chr11-119101945-119102445;chr11-119107050-119107550                     </td><td> 2.480134e-14</td><td> 3.478512e-13</td></tr>\n",
       "\t<tr><td>MORN3   </td><td>ZNF263 </td><td> 0.1730188735</td><td>4</td><td>3</td><td>3</td><td>chr12-121672401-121672901                                                                                                   </td><td> 4.654712e-03</td><td> 2.405455e-02</td></tr>\n",
       "\t<tr><td>TOP1MT  </td><td>ZNF263 </td><td> 0.0339685661</td><td>3</td><td>3</td><td>3</td><td>chr8-143246465-143246965                                                                                                    </td><td> 7.075552e-05</td><td> 4.825664e-04</td></tr>\n",
       "\t<tr><td>MORN3   </td><td>ZNF274 </td><td> 0.4962315004</td><td>4</td><td>1</td><td>3</td><td>chr12-121651385-121651885                                                                                                   </td><td> 6.034788e-06</td><td> 4.671111e-05</td></tr>\n",
       "\t<tr><td>SLC15A3 </td><td>ZNF282 </td><td> 1.4494766817</td><td>4</td><td>1</td><td>2</td><td>chr11-60906385-60906885;chr11-60910901-60911401;chr11-60978146-60978646                                                     </td><td> 1.162904e-61</td><td> 8.140331e-60</td></tr>\n",
       "\t<tr><td>TM9SF1  </td><td>ZNF341 </td><td> 0.2128660511</td><td>5</td><td>1</td><td>3</td><td>chr14-24115146-24115646;chr14-24164931-24165431;chr14-24188154-24188654                                                     </td><td> 3.531118e-18</td><td> 6.082687e-17</td></tr>\n",
       "\t<tr><td>VWA1    </td><td>ZNF382 </td><td> 1.3110463427</td><td>3</td><td>1</td><td>2</td><td>chr1-1341955-1342455                                                                                                        </td><td> 3.506171e-19</td><td> 6.307830e-18</td></tr>\n",
       "\t<tr><td>NAXE    </td><td>ZNF384 </td><td> 0.1316343652</td><td>1</td><td>1</td><td>2</td><td>chr1-156501753-156502253                                                                                                    </td><td> 5.897128e-03</td><td> 2.987361e-02</td></tr>\n",
       "\t<tr><td>DAXX    </td><td>ZNF449 </td><td>-0.0091632543</td><td>8</td><td>1</td><td>3</td><td>chr6-33297255-33297755;chr6-33323508-33324008                                                                               </td><td> 4.424444e-09</td><td> 4.430197e-08</td></tr>\n",
       "\t<tr><td>CBX8    </td><td>ZNF460 </td><td> 1.3062750972</td><td>1</td><td>4</td><td>1</td><td>chr17-79810041-79810541                                                                                                     </td><td>4.274129e-162</td><td>2.194053e-159</td></tr>\n",
       "\t<tr><td>ERCC2   </td><td>ZNF460 </td><td> 0.4955387211</td><td>2</td><td>4</td><td>1</td><td>chr19-45450199-45450699;chr19-45467434-45467934                                                                             </td><td> 3.431580e-36</td><td> 1.129195e-34</td></tr>\n",
       "\t<tr><td>FN3K    </td><td>ZNF460 </td><td> 1.4443375251</td><td>1</td><td>4</td><td>1</td><td>chr17-82648790-82649290                                                                                                     </td><td> 1.092093e-23</td><td> 2.297573e-22</td></tr>\n",
       "\t<tr><td>HEY1    </td><td>ZNF460 </td><td> 0.6340373533</td><td>1</td><td>4</td><td>1</td><td>chr8-79767703-79768203                                                                                                      </td><td> 5.629898e-43</td><td> 2.269645e-41</td></tr>\n",
       "\t<tr><td>TMEM223 </td><td>ZNF682 </td><td> 0.1438188379</td><td>4</td><td>1</td><td>3</td><td>chr11-62805262-62805762;chr11-62831890-62832390                                                                             </td><td> 1.375070e-23</td><td> 2.869387e-22</td></tr>\n",
       "\t<tr><td>PTK7    </td><td>ZNF684 </td><td>-0.0009780898</td><td>3</td><td>1</td><td>2</td><td>chr6-43013818-43014318;chr6-43014425-43014925                                                                               </td><td> 4.279343e-42</td><td> 1.655826e-40</td></tr>\n",
       "\t<tr><td>CREB3L4 </td><td>ZNF75D </td><td> 0.4131517495</td><td>3</td><td>1</td><td>2</td><td>chr1-153921559-153922059;chr1-153944302-153944802                                                                           </td><td> 3.859810e-67</td><td> 3.084533e-65</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 148 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " target & tf & estimate & n\\_regions & n\\_genes & n\\_tfs & regions & pval & padj\\\\\n",
       " <chr> & <chr> & <dbl> & <int> & <int> & <int> & <chr> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t HMBS      & CREB3 & 0.502456918 & 1 & 1 & 1 & chr11-119057030-119057530                                            &  1.413769e-16 &  2.239922e-15\\\\\n",
       "\t TNNT3     & E2F6  & 0.150808737 & 2 & 1 & 2 & chr11-1839764-1840264                                                &  7.670730e-35 &  2.391280e-33\\\\\n",
       "\t BBC3      & EGR2  & 0.221948001 & 3 & 1 & 3 & chr19-47274053-47274553                                              &  2.391756e-05 &  1.727628e-04\\\\\n",
       "\t KEAP1     & EGR3  & 1.008195189 & 3 & 1 & 2 & chr19-10478848-10479348;chr19-10602148-10602648                      &  9.042581e-10 &  9.564268e-09\\\\\n",
       "\t GZMB      & EHF   & 0.644146878 & 4 & 1 & 3 & chr14-24695188-24695688                                              &  2.421297e-10 &  2.663427e-09\\\\\n",
       "\t TOP1MT    & ELF3  & 0.415747555 & 3 & 1 & 3 & chr8-143281493-143281993                                             &  3.595062e-18 &  6.179013e-17\\\\\n",
       "\t LINC01569 & ELK3  & 0.013216023 & 3 & 2 & 2 & chr16-4307497-4307997                                                &  8.634882e-04 &  5.052324e-03\\\\\n",
       "\t SPSB2     & ELK3  & 0.003256462 & 3 & 2 & 2 & chr12-6943749-6944249;chr12-6970506-6971006                          &  6.425397e-08 &  5.882944e-07\\\\\n",
       "\t FCN3      & ELK4  & 0.061526038 & 2 & 3 & 2 & chr1-27321907-27322407                                               &  3.623591e-05 &  2.557763e-04\\\\\n",
       "\t SNHG15    & ELK4  & 0.256699658 & 4 & 3 & 3 & chr7-44999794-45000294;chr7-45000351-45000851                        &  5.909617e-04 &  3.546691e-03\\\\\n",
       "\t STAP2     & ELK4  & 0.035338163 & 2 & 3 & 2 & chr19-4400389-4400889                                                &  3.732396e-05 &  2.631818e-04\\\\\n",
       "\t STOML2    & ESR1  & 0.182048769 & 3 & 1 & 2 & chr9-35080311-35080811;chr9-35095760-35096260                        &  8.314858e-34 &  2.462477e-32\\\\\n",
       "\t CCR7      & ESR2  & 0.429512770 & 4 & 3 & 3 & chr17-40582251-40582751                                              &  2.734040e-11 &  3.209162e-10\\\\\n",
       "\t FCGR3A    & ESR2  & 0.341330263 & 5 & 3 & 5 & chr1-161545274-161545774                                             &  4.870036e-14 &  6.744474e-13\\\\\n",
       "\t NLGN2     & ESR2  & 0.110788987 & 3 & 3 & 1 & chr17-7313890-7314390;chr17-7351839-7352339;chr17-7352410-7352910    &  1.293830e-31 &  3.596567e-30\\\\\n",
       "\t SYTL3     & ETS1  & 0.183452614 & 1 & 1 & 1 & chr6-158643111-158643611                                             &  3.694966e-03 &  1.942064e-02\\\\\n",
       "\t PTCSC2    & ETS2  & 0.498485796 & 1 & 1 & 1 & chr9-97904628-97905128                                               &  6.605345e-51 &  3.413501e-49\\\\\n",
       "\t DAXX      & ETV4  & 0.491687344 & 8 & 1 & 3 & chr6-33249651-33250151                                               &  4.476906e-14 &  6.222414e-13\\\\\n",
       "\t AFMID     & ETV5  & 0.212461085 & 3 & 1 & 3 & chr17-78130482-78130982                                              &  2.535156e-05 &  1.820961e-04\\\\\n",
       "\t CCR7      & ETV6  & 0.130363935 & 4 & 1 & 3 & chr17-40565166-40565666;chr17-40647845-40648345                      &  1.798356e-03 &  9.912200e-03\\\\\n",
       "\t SERINC5   & FOXP1 & 0.314644062 & 3 & 1 & 2 & chr5-80186588-80187088                                               &  5.482104e-04 &  3.302989e-03\\\\\n",
       "\t CTU2      & FOXP2 & 0.064735297 & 8 & 1 & 2 & chr16-88618912-88619412;chr16-88710799-88711299                      &  1.936689e-21 &  3.784900e-20\\\\\n",
       "\t ATP6V1G2  & GABPA & 0.188905241 & 5 & 1 & 4 & chr6-31541808-31542308;chr6-31569306-31569806;chr6-31576740-31577240 &  2.638782e-04 &  1.657310e-03\\\\\n",
       "\t BBC3      & GFI1  & 0.492520606 & 3 & 2 & 3 & chr19-47261875-47262375                                              &  2.135358e-12 &  2.722228e-11\\\\\n",
       "\t FBXO2     & GFI1  & 0.358505134 & 1 & 2 & 1 & chr1-11708563-11709063                                               &  2.324456e-30 &  6.150622e-29\\\\\n",
       "\t AFMID     & GLI2  & 0.404393915 & 3 & 1 & 3 & chr17-78120864-78121364                                              &  3.210560e-34 &  9.732799e-33\\\\\n",
       "\t ZNF843    & GLIS2 & 0.321331869 & 4 & 1 & 4 & chr16-31458886-31459386                                              & 3.321651e-191 & 1.967439e-188\\\\\n",
       "\t BBC3      & HIC2  & 0.666970140 & 3 & 1 & 3 & chr19-47129956-47130456                                              &  2.760745e-15 &  4.088026e-14\\\\\n",
       "\t CREB3L4   & HOXB2 & 0.498664799 & 3 & 1 & 2 & chr1-153945788-153946288                                             &  7.034193e-72 &  6.298056e-70\\\\\n",
       "\t FCGR3A    & HOXB4 & 0.828054849 & 5 & 1 & 5 & chr1-161550372-161550872                                             &  6.479635e-08 &  5.918528e-07\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t MORN3    & ZBTB14  &  0.8318287039 & 4 & 1 & 3 & chr12-121560906-121561406;chr12-121658845-121659345                                                                          &  1.133754e-13 &  1.542387e-12\\\\\n",
       "\t DALRD3   & ZBTB18  &  1.0167926147 & 2 & 2 & 2 & chr3-49021835-49022335                                                                                                       &  1.715073e-05 &  1.260120e-04\\\\\n",
       "\t FCN3     & ZBTB18  &  0.6407947401 & 2 & 2 & 2 & chr1-27343258-27343758                                                                                                       &  1.338906e-38 &  4.863006e-37\\\\\n",
       "\t GSTO2    & ZBTB33  &  0.4757434981 & 1 & 1 & 1 & chr10-104268851-104269351                                                                                                    &  7.522957e-40 &  2.825696e-38\\\\\n",
       "\t DALRD3   & ZBTB6   &  0.9895490015 & 2 & 2 & 2 & chr3-48989626-48990126                                                                                                       &  1.045595e-23 &  2.205776e-22\\\\\n",
       "\t VAMP5    & ZBTB6   &  0.4881793017 & 2 & 2 & 2 & chr2-85561906-85562406                                                                                                       &  2.940189e-11 &  3.440647e-10\\\\\n",
       "\t DAXX     & ZKSCAN5 &  0.9722396972 & 8 & 2 & 3 & chr6-33271742-33272242;chr6-33279915-33280415;chr6-33298424-33298924;chr6-33322893-33323393;chr6-33425773-33426273           &  1.512623e-17 &  2.488718e-16\\\\\n",
       "\t KLF16    & ZKSCAN5 &  0.5000000000 & 4 & 2 & 2 & chr19-1848488-1848988;chr19-1873393-1873893;chr19-1882818-1883318                                                            &  0.000000e+00 &  0.000000e+00\\\\\n",
       "\t C16orf74 & ZNF135  &  0.3696168578 & 2 & 2 & 2 & chr16-85611569-85612069                                                                                                      &  2.278164e-27 &  5.447784e-26\\\\\n",
       "\t TNNT3    & ZNF135  &  0.9969535196 & 2 & 2 & 2 & chr11-1878260-1878760                                                                                                        &  1.632778e-29 &  4.190796e-28\\\\\n",
       "\t TOP1MT   & ZNF143  &  0.5171160910 & 3 & 1 & 3 & chr8-143246465-143246965;chr8-143290355-143290855                                                                            &  8.969845e-51 &  4.574027e-49\\\\\n",
       "\t CLIC3    & ZNF148  &  0.7189429748 & 7 & 1 & 2 & chr9-136949884-136950384;chr9-136958911-136959411;chr9-136960139-136960639;chr9-136960745-136961245;chr9-136970157-136970657 &  1.672711e-10 &  1.861253e-09\\\\\n",
       "\t FDXR     & ZNF16   &  0.7657769925 & 2 & 2 & 2 & chr17-74871275-74871775                                                                                                      &  7.373991e-34 &  2.192268e-32\\\\\n",
       "\t IRGQ     & ZNF16   &  1.2608986329 & 3 & 2 & 3 & chr19-43504488-43504988                                                                                                      &  4.920331e-30 &  1.279951e-28\\\\\n",
       "\t C2CD2L   & ZNF263  &  0.8435557001 & 5 & 3 & 3 & chr11-119057030-119057530;chr11-119100173-119100673;chr11-119101945-119102445;chr11-119107050-119107550                      &  2.480134e-14 &  3.478512e-13\\\\\n",
       "\t MORN3    & ZNF263  &  0.1730188735 & 4 & 3 & 3 & chr12-121672401-121672901                                                                                                    &  4.654712e-03 &  2.405455e-02\\\\\n",
       "\t TOP1MT   & ZNF263  &  0.0339685661 & 3 & 3 & 3 & chr8-143246465-143246965                                                                                                     &  7.075552e-05 &  4.825664e-04\\\\\n",
       "\t MORN3    & ZNF274  &  0.4962315004 & 4 & 1 & 3 & chr12-121651385-121651885                                                                                                    &  6.034788e-06 &  4.671111e-05\\\\\n",
       "\t SLC15A3  & ZNF282  &  1.4494766817 & 4 & 1 & 2 & chr11-60906385-60906885;chr11-60910901-60911401;chr11-60978146-60978646                                                      &  1.162904e-61 &  8.140331e-60\\\\\n",
       "\t TM9SF1   & ZNF341  &  0.2128660511 & 5 & 1 & 3 & chr14-24115146-24115646;chr14-24164931-24165431;chr14-24188154-24188654                                                      &  3.531118e-18 &  6.082687e-17\\\\\n",
       "\t VWA1     & ZNF382  &  1.3110463427 & 3 & 1 & 2 & chr1-1341955-1342455                                                                                                         &  3.506171e-19 &  6.307830e-18\\\\\n",
       "\t NAXE     & ZNF384  &  0.1316343652 & 1 & 1 & 2 & chr1-156501753-156502253                                                                                                     &  5.897128e-03 &  2.987361e-02\\\\\n",
       "\t DAXX     & ZNF449  & -0.0091632543 & 8 & 1 & 3 & chr6-33297255-33297755;chr6-33323508-33324008                                                                                &  4.424444e-09 &  4.430197e-08\\\\\n",
       "\t CBX8     & ZNF460  &  1.3062750972 & 1 & 4 & 1 & chr17-79810041-79810541                                                                                                      & 4.274129e-162 & 2.194053e-159\\\\\n",
       "\t ERCC2    & ZNF460  &  0.4955387211 & 2 & 4 & 1 & chr19-45450199-45450699;chr19-45467434-45467934                                                                              &  3.431580e-36 &  1.129195e-34\\\\\n",
       "\t FN3K     & ZNF460  &  1.4443375251 & 1 & 4 & 1 & chr17-82648790-82649290                                                                                                      &  1.092093e-23 &  2.297573e-22\\\\\n",
       "\t HEY1     & ZNF460  &  0.6340373533 & 1 & 4 & 1 & chr8-79767703-79768203                                                                                                       &  5.629898e-43 &  2.269645e-41\\\\\n",
       "\t TMEM223  & ZNF682  &  0.1438188379 & 4 & 1 & 3 & chr11-62805262-62805762;chr11-62831890-62832390                                                                              &  1.375070e-23 &  2.869387e-22\\\\\n",
       "\t PTK7     & ZNF684  & -0.0009780898 & 3 & 1 & 2 & chr6-43013818-43014318;chr6-43014425-43014925                                                                                &  4.279343e-42 &  1.655826e-40\\\\\n",
       "\t CREB3L4  & ZNF75D  &  0.4131517495 & 3 & 1 & 2 & chr1-153921559-153922059;chr1-153944302-153944802                                                                            &  3.859810e-67 &  3.084533e-65\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 148 × 9\n",
       "\n",
       "| target &lt;chr&gt; | tf &lt;chr&gt; | estimate &lt;dbl&gt; | n_regions &lt;int&gt; | n_genes &lt;int&gt; | n_tfs &lt;int&gt; | regions &lt;chr&gt; | pval &lt;dbl&gt; | padj &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| HMBS      | CREB3 | 0.502456918 | 1 | 1 | 1 | chr11-119057030-119057530                                            |  1.413769e-16 |  2.239922e-15 |\n",
       "| TNNT3     | E2F6  | 0.150808737 | 2 | 1 | 2 | chr11-1839764-1840264                                                |  7.670730e-35 |  2.391280e-33 |\n",
       "| BBC3      | EGR2  | 0.221948001 | 3 | 1 | 3 | chr19-47274053-47274553                                              |  2.391756e-05 |  1.727628e-04 |\n",
       "| KEAP1     | EGR3  | 1.008195189 | 3 | 1 | 2 | chr19-10478848-10479348;chr19-10602148-10602648                      |  9.042581e-10 |  9.564268e-09 |\n",
       "| GZMB      | EHF   | 0.644146878 | 4 | 1 | 3 | chr14-24695188-24695688                                              |  2.421297e-10 |  2.663427e-09 |\n",
       "| TOP1MT    | ELF3  | 0.415747555 | 3 | 1 | 3 | chr8-143281493-143281993                                             |  3.595062e-18 |  6.179013e-17 |\n",
       "| LINC01569 | ELK3  | 0.013216023 | 3 | 2 | 2 | chr16-4307497-4307997                                                |  8.634882e-04 |  5.052324e-03 |\n",
       "| SPSB2     | ELK3  | 0.003256462 | 3 | 2 | 2 | chr12-6943749-6944249;chr12-6970506-6971006                          |  6.425397e-08 |  5.882944e-07 |\n",
       "| FCN3      | ELK4  | 0.061526038 | 2 | 3 | 2 | chr1-27321907-27322407                                               |  3.623591e-05 |  2.557763e-04 |\n",
       "| SNHG15    | ELK4  | 0.256699658 | 4 | 3 | 3 | chr7-44999794-45000294;chr7-45000351-45000851                        |  5.909617e-04 |  3.546691e-03 |\n",
       "| STAP2     | ELK4  | 0.035338163 | 2 | 3 | 2 | chr19-4400389-4400889                                                |  3.732396e-05 |  2.631818e-04 |\n",
       "| STOML2    | ESR1  | 0.182048769 | 3 | 1 | 2 | chr9-35080311-35080811;chr9-35095760-35096260                        |  8.314858e-34 |  2.462477e-32 |\n",
       "| CCR7      | ESR2  | 0.429512770 | 4 | 3 | 3 | chr17-40582251-40582751                                              |  2.734040e-11 |  3.209162e-10 |\n",
       "| FCGR3A    | ESR2  | 0.341330263 | 5 | 3 | 5 | chr1-161545274-161545774                                             |  4.870036e-14 |  6.744474e-13 |\n",
       "| NLGN2     | ESR2  | 0.110788987 | 3 | 3 | 1 | chr17-7313890-7314390;chr17-7351839-7352339;chr17-7352410-7352910    |  1.293830e-31 |  3.596567e-30 |\n",
       "| SYTL3     | ETS1  | 0.183452614 | 1 | 1 | 1 | chr6-158643111-158643611                                             |  3.694966e-03 |  1.942064e-02 |\n",
       "| PTCSC2    | ETS2  | 0.498485796 | 1 | 1 | 1 | chr9-97904628-97905128                                               |  6.605345e-51 |  3.413501e-49 |\n",
       "| DAXX      | ETV4  | 0.491687344 | 8 | 1 | 3 | chr6-33249651-33250151                                               |  4.476906e-14 |  6.222414e-13 |\n",
       "| AFMID     | ETV5  | 0.212461085 | 3 | 1 | 3 | chr17-78130482-78130982                                              |  2.535156e-05 |  1.820961e-04 |\n",
       "| CCR7      | ETV6  | 0.130363935 | 4 | 1 | 3 | chr17-40565166-40565666;chr17-40647845-40648345                      |  1.798356e-03 |  9.912200e-03 |\n",
       "| SERINC5   | FOXP1 | 0.314644062 | 3 | 1 | 2 | chr5-80186588-80187088                                               |  5.482104e-04 |  3.302989e-03 |\n",
       "| CTU2      | FOXP2 | 0.064735297 | 8 | 1 | 2 | chr16-88618912-88619412;chr16-88710799-88711299                      |  1.936689e-21 |  3.784900e-20 |\n",
       "| ATP6V1G2  | GABPA | 0.188905241 | 5 | 1 | 4 | chr6-31541808-31542308;chr6-31569306-31569806;chr6-31576740-31577240 |  2.638782e-04 |  1.657310e-03 |\n",
       "| BBC3      | GFI1  | 0.492520606 | 3 | 2 | 3 | chr19-47261875-47262375                                              |  2.135358e-12 |  2.722228e-11 |\n",
       "| FBXO2     | GFI1  | 0.358505134 | 1 | 2 | 1 | chr1-11708563-11709063                                               |  2.324456e-30 |  6.150622e-29 |\n",
       "| AFMID     | GLI2  | 0.404393915 | 3 | 1 | 3 | chr17-78120864-78121364                                              |  3.210560e-34 |  9.732799e-33 |\n",
       "| ZNF843    | GLIS2 | 0.321331869 | 4 | 1 | 4 | chr16-31458886-31459386                                              | 3.321651e-191 | 1.967439e-188 |\n",
       "| BBC3      | HIC2  | 0.666970140 | 3 | 1 | 3 | chr19-47129956-47130456                                              |  2.760745e-15 |  4.088026e-14 |\n",
       "| CREB3L4   | HOXB2 | 0.498664799 | 3 | 1 | 2 | chr1-153945788-153946288                                             |  7.034193e-72 |  6.298056e-70 |\n",
       "| FCGR3A    | HOXB4 | 0.828054849 | 5 | 1 | 5 | chr1-161550372-161550872                                             |  6.479635e-08 |  5.918528e-07 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| MORN3    | ZBTB14  |  0.8318287039 | 4 | 1 | 3 | chr12-121560906-121561406;chr12-121658845-121659345                                                                          |  1.133754e-13 |  1.542387e-12 |\n",
       "| DALRD3   | ZBTB18  |  1.0167926147 | 2 | 2 | 2 | chr3-49021835-49022335                                                                                                       |  1.715073e-05 |  1.260120e-04 |\n",
       "| FCN3     | ZBTB18  |  0.6407947401 | 2 | 2 | 2 | chr1-27343258-27343758                                                                                                       |  1.338906e-38 |  4.863006e-37 |\n",
       "| GSTO2    | ZBTB33  |  0.4757434981 | 1 | 1 | 1 | chr10-104268851-104269351                                                                                                    |  7.522957e-40 |  2.825696e-38 |\n",
       "| DALRD3   | ZBTB6   |  0.9895490015 | 2 | 2 | 2 | chr3-48989626-48990126                                                                                                       |  1.045595e-23 |  2.205776e-22 |\n",
       "| VAMP5    | ZBTB6   |  0.4881793017 | 2 | 2 | 2 | chr2-85561906-85562406                                                                                                       |  2.940189e-11 |  3.440647e-10 |\n",
       "| DAXX     | ZKSCAN5 |  0.9722396972 | 8 | 2 | 3 | chr6-33271742-33272242;chr6-33279915-33280415;chr6-33298424-33298924;chr6-33322893-33323393;chr6-33425773-33426273           |  1.512623e-17 |  2.488718e-16 |\n",
       "| KLF16    | ZKSCAN5 |  0.5000000000 | 4 | 2 | 2 | chr19-1848488-1848988;chr19-1873393-1873893;chr19-1882818-1883318                                                            |  0.000000e+00 |  0.000000e+00 |\n",
       "| C16orf74 | ZNF135  |  0.3696168578 | 2 | 2 | 2 | chr16-85611569-85612069                                                                                                      |  2.278164e-27 |  5.447784e-26 |\n",
       "| TNNT3    | ZNF135  |  0.9969535196 | 2 | 2 | 2 | chr11-1878260-1878760                                                                                                        |  1.632778e-29 |  4.190796e-28 |\n",
       "| TOP1MT   | ZNF143  |  0.5171160910 | 3 | 1 | 3 | chr8-143246465-143246965;chr8-143290355-143290855                                                                            |  8.969845e-51 |  4.574027e-49 |\n",
       "| CLIC3    | ZNF148  |  0.7189429748 | 7 | 1 | 2 | chr9-136949884-136950384;chr9-136958911-136959411;chr9-136960139-136960639;chr9-136960745-136961245;chr9-136970157-136970657 |  1.672711e-10 |  1.861253e-09 |\n",
       "| FDXR     | ZNF16   |  0.7657769925 | 2 | 2 | 2 | chr17-74871275-74871775                                                                                                      |  7.373991e-34 |  2.192268e-32 |\n",
       "| IRGQ     | ZNF16   |  1.2608986329 | 3 | 2 | 3 | chr19-43504488-43504988                                                                                                      |  4.920331e-30 |  1.279951e-28 |\n",
       "| C2CD2L   | ZNF263  |  0.8435557001 | 5 | 3 | 3 | chr11-119057030-119057530;chr11-119100173-119100673;chr11-119101945-119102445;chr11-119107050-119107550                      |  2.480134e-14 |  3.478512e-13 |\n",
       "| MORN3    | ZNF263  |  0.1730188735 | 4 | 3 | 3 | chr12-121672401-121672901                                                                                                    |  4.654712e-03 |  2.405455e-02 |\n",
       "| TOP1MT   | ZNF263  |  0.0339685661 | 3 | 3 | 3 | chr8-143246465-143246965                                                                                                     |  7.075552e-05 |  4.825664e-04 |\n",
       "| MORN3    | ZNF274  |  0.4962315004 | 4 | 1 | 3 | chr12-121651385-121651885                                                                                                    |  6.034788e-06 |  4.671111e-05 |\n",
       "| SLC15A3  | ZNF282  |  1.4494766817 | 4 | 1 | 2 | chr11-60906385-60906885;chr11-60910901-60911401;chr11-60978146-60978646                                                      |  1.162904e-61 |  8.140331e-60 |\n",
       "| TM9SF1   | ZNF341  |  0.2128660511 | 5 | 1 | 3 | chr14-24115146-24115646;chr14-24164931-24165431;chr14-24188154-24188654                                                      |  3.531118e-18 |  6.082687e-17 |\n",
       "| VWA1     | ZNF382  |  1.3110463427 | 3 | 1 | 2 | chr1-1341955-1342455                                                                                                         |  3.506171e-19 |  6.307830e-18 |\n",
       "| NAXE     | ZNF384  |  0.1316343652 | 1 | 1 | 2 | chr1-156501753-156502253                                                                                                     |  5.897128e-03 |  2.987361e-02 |\n",
       "| DAXX     | ZNF449  | -0.0091632543 | 8 | 1 | 3 | chr6-33297255-33297755;chr6-33323508-33324008                                                                                |  4.424444e-09 |  4.430197e-08 |\n",
       "| CBX8     | ZNF460  |  1.3062750972 | 1 | 4 | 1 | chr17-79810041-79810541                                                                                                      | 4.274129e-162 | 2.194053e-159 |\n",
       "| ERCC2    | ZNF460  |  0.4955387211 | 2 | 4 | 1 | chr19-45450199-45450699;chr19-45467434-45467934                                                                              |  3.431580e-36 |  1.129195e-34 |\n",
       "| FN3K     | ZNF460  |  1.4443375251 | 1 | 4 | 1 | chr17-82648790-82649290                                                                                                      |  1.092093e-23 |  2.297573e-22 |\n",
       "| HEY1     | ZNF460  |  0.6340373533 | 1 | 4 | 1 | chr8-79767703-79768203                                                                                                       |  5.629898e-43 |  2.269645e-41 |\n",
       "| TMEM223  | ZNF682  |  0.1438188379 | 4 | 1 | 3 | chr11-62805262-62805762;chr11-62831890-62832390                                                                              |  1.375070e-23 |  2.869387e-22 |\n",
       "| PTK7     | ZNF684  | -0.0009780898 | 3 | 1 | 2 | chr6-43013818-43014318;chr6-43014425-43014925                                                                                |  4.279343e-42 |  1.655826e-40 |\n",
       "| CREB3L4  | ZNF75D  |  0.4131517495 | 3 | 1 | 2 | chr1-153921559-153922059;chr1-153944302-153944802                                                                            |  3.859810e-67 |  3.084533e-65 |\n",
       "\n"
      ],
      "text/plain": [
       "    target    tf      estimate      n_regions n_genes n_tfs\n",
       "1   HMBS      CREB3   0.502456918   1         1       1    \n",
       "2   TNNT3     E2F6    0.150808737   2         1       2    \n",
       "3   BBC3      EGR2    0.221948001   3         1       3    \n",
       "4   KEAP1     EGR3    1.008195189   3         1       2    \n",
       "5   GZMB      EHF     0.644146878   4         1       3    \n",
       "6   TOP1MT    ELF3    0.415747555   3         1       3    \n",
       "7   LINC01569 ELK3    0.013216023   3         2       2    \n",
       "8   SPSB2     ELK3    0.003256462   3         2       2    \n",
       "9   FCN3      ELK4    0.061526038   2         3       2    \n",
       "10  SNHG15    ELK4    0.256699658   4         3       3    \n",
       "11  STAP2     ELK4    0.035338163   2         3       2    \n",
       "12  STOML2    ESR1    0.182048769   3         1       2    \n",
       "13  CCR7      ESR2    0.429512770   4         3       3    \n",
       "14  FCGR3A    ESR2    0.341330263   5         3       5    \n",
       "15  NLGN2     ESR2    0.110788987   3         3       1    \n",
       "16  SYTL3     ETS1    0.183452614   1         1       1    \n",
       "17  PTCSC2    ETS2    0.498485796   1         1       1    \n",
       "18  DAXX      ETV4    0.491687344   8         1       3    \n",
       "19  AFMID     ETV5    0.212461085   3         1       3    \n",
       "20  CCR7      ETV6    0.130363935   4         1       3    \n",
       "21  SERINC5   FOXP1   0.314644062   3         1       2    \n",
       "22  CTU2      FOXP2   0.064735297   8         1       2    \n",
       "23  ATP6V1G2  GABPA   0.188905241   5         1       4    \n",
       "24  BBC3      GFI1    0.492520606   3         2       3    \n",
       "25  FBXO2     GFI1    0.358505134   1         2       1    \n",
       "26  AFMID     GLI2    0.404393915   3         1       3    \n",
       "27  ZNF843    GLIS2   0.321331869   4         1       4    \n",
       "28  BBC3      HIC2    0.666970140   3         1       3    \n",
       "29  CREB3L4   HOXB2   0.498664799   3         1       2    \n",
       "30  FCGR3A    HOXB4   0.828054849   5         1       5    \n",
       "⋮   ⋮         ⋮       ⋮             ⋮         ⋮       ⋮    \n",
       "119 MORN3     ZBTB14   0.8318287039 4         1       3    \n",
       "120 DALRD3    ZBTB18   1.0167926147 2         2       2    \n",
       "121 FCN3      ZBTB18   0.6407947401 2         2       2    \n",
       "122 GSTO2     ZBTB33   0.4757434981 1         1       1    \n",
       "123 DALRD3    ZBTB6    0.9895490015 2         2       2    \n",
       "124 VAMP5     ZBTB6    0.4881793017 2         2       2    \n",
       "125 DAXX      ZKSCAN5  0.9722396972 8         2       3    \n",
       "126 KLF16     ZKSCAN5  0.5000000000 4         2       2    \n",
       "127 C16orf74  ZNF135   0.3696168578 2         2       2    \n",
       "128 TNNT3     ZNF135   0.9969535196 2         2       2    \n",
       "129 TOP1MT    ZNF143   0.5171160910 3         1       3    \n",
       "130 CLIC3     ZNF148   0.7189429748 7         1       2    \n",
       "131 FDXR      ZNF16    0.7657769925 2         2       2    \n",
       "132 IRGQ      ZNF16    1.2608986329 3         2       3    \n",
       "133 C2CD2L    ZNF263   0.8435557001 5         3       3    \n",
       "134 MORN3     ZNF263   0.1730188735 4         3       3    \n",
       "135 TOP1MT    ZNF263   0.0339685661 3         3       3    \n",
       "136 MORN3     ZNF274   0.4962315004 4         1       3    \n",
       "137 SLC15A3   ZNF282   1.4494766817 4         1       2    \n",
       "138 TM9SF1    ZNF341   0.2128660511 5         1       3    \n",
       "139 VWA1      ZNF382   1.3110463427 3         1       2    \n",
       "140 NAXE      ZNF384   0.1316343652 1         1       2    \n",
       "141 DAXX      ZNF449  -0.0091632543 8         1       3    \n",
       "142 CBX8      ZNF460   1.3062750972 1         4       1    \n",
       "143 ERCC2     ZNF460   0.4955387211 2         4       1    \n",
       "144 FN3K      ZNF460   1.4443375251 1         4       1    \n",
       "145 HEY1      ZNF460   0.6340373533 1         4       1    \n",
       "146 TMEM223   ZNF682   0.1438188379 4         1       3    \n",
       "147 PTK7      ZNF684  -0.0009780898 3         1       2    \n",
       "148 CREB3L4   ZNF75D   0.4131517495 3         1       2    \n",
       "    regions                                                                                                                     \n",
       "1   chr11-119057030-119057530                                                                                                   \n",
       "2   chr11-1839764-1840264                                                                                                       \n",
       "3   chr19-47274053-47274553                                                                                                     \n",
       "4   chr19-10478848-10479348;chr19-10602148-10602648                                                                             \n",
       "5   chr14-24695188-24695688                                                                                                     \n",
       "6   chr8-143281493-143281993                                                                                                    \n",
       "7   chr16-4307497-4307997                                                                                                       \n",
       "8   chr12-6943749-6944249;chr12-6970506-6971006                                                                                 \n",
       "9   chr1-27321907-27322407                                                                                                      \n",
       "10  chr7-44999794-45000294;chr7-45000351-45000851                                                                               \n",
       "11  chr19-4400389-4400889                                                                                                       \n",
       "12  chr9-35080311-35080811;chr9-35095760-35096260                                                                               \n",
       "13  chr17-40582251-40582751                                                                                                     \n",
       "14  chr1-161545274-161545774                                                                                                    \n",
       "15  chr17-7313890-7314390;chr17-7351839-7352339;chr17-7352410-7352910                                                           \n",
       "16  chr6-158643111-158643611                                                                                                    \n",
       "17  chr9-97904628-97905128                                                                                                      \n",
       "18  chr6-33249651-33250151                                                                                                      \n",
       "19  chr17-78130482-78130982                                                                                                     \n",
       "20  chr17-40565166-40565666;chr17-40647845-40648345                                                                             \n",
       "21  chr5-80186588-80187088                                                                                                      \n",
       "22  chr16-88618912-88619412;chr16-88710799-88711299                                                                             \n",
       "23  chr6-31541808-31542308;chr6-31569306-31569806;chr6-31576740-31577240                                                        \n",
       "24  chr19-47261875-47262375                                                                                                     \n",
       "25  chr1-11708563-11709063                                                                                                      \n",
       "26  chr17-78120864-78121364                                                                                                     \n",
       "27  chr16-31458886-31459386                                                                                                     \n",
       "28  chr19-47129956-47130456                                                                                                     \n",
       "29  chr1-153945788-153946288                                                                                                    \n",
       "30  chr1-161550372-161550872                                                                                                    \n",
       "⋮   ⋮                                                                                                                           \n",
       "119 chr12-121560906-121561406;chr12-121658845-121659345                                                                         \n",
       "120 chr3-49021835-49022335                                                                                                      \n",
       "121 chr1-27343258-27343758                                                                                                      \n",
       "122 chr10-104268851-104269351                                                                                                   \n",
       "123 chr3-48989626-48990126                                                                                                      \n",
       "124 chr2-85561906-85562406                                                                                                      \n",
       "125 chr6-33271742-33272242;chr6-33279915-33280415;chr6-33298424-33298924;chr6-33322893-33323393;chr6-33425773-33426273          \n",
       "126 chr19-1848488-1848988;chr19-1873393-1873893;chr19-1882818-1883318                                                           \n",
       "127 chr16-85611569-85612069                                                                                                     \n",
       "128 chr11-1878260-1878760                                                                                                       \n",
       "129 chr8-143246465-143246965;chr8-143290355-143290855                                                                           \n",
       "130 chr9-136949884-136950384;chr9-136958911-136959411;chr9-136960139-136960639;chr9-136960745-136961245;chr9-136970157-136970657\n",
       "131 chr17-74871275-74871775                                                                                                     \n",
       "132 chr19-43504488-43504988                                                                                                     \n",
       "133 chr11-119057030-119057530;chr11-119100173-119100673;chr11-119101945-119102445;chr11-119107050-119107550                     \n",
       "134 chr12-121672401-121672901                                                                                                   \n",
       "135 chr8-143246465-143246965                                                                                                    \n",
       "136 chr12-121651385-121651885                                                                                                   \n",
       "137 chr11-60906385-60906885;chr11-60910901-60911401;chr11-60978146-60978646                                                     \n",
       "138 chr14-24115146-24115646;chr14-24164931-24165431;chr14-24188154-24188654                                                     \n",
       "139 chr1-1341955-1342455                                                                                                        \n",
       "140 chr1-156501753-156502253                                                                                                    \n",
       "141 chr6-33297255-33297755;chr6-33323508-33324008                                                                               \n",
       "142 chr17-79810041-79810541                                                                                                     \n",
       "143 chr19-45450199-45450699;chr19-45467434-45467934                                                                             \n",
       "144 chr17-82648790-82649290                                                                                                     \n",
       "145 chr8-79767703-79768203                                                                                                      \n",
       "146 chr11-62805262-62805762;chr11-62831890-62832390                                                                             \n",
       "147 chr6-43013818-43014318;chr6-43014425-43014925                                                                               \n",
       "148 chr1-153921559-153922059;chr1-153944302-153944802                                                                           \n",
       "    pval          padj         \n",
       "1    1.413769e-16  2.239922e-15\n",
       "2    7.670730e-35  2.391280e-33\n",
       "3    2.391756e-05  1.727628e-04\n",
       "4    9.042581e-10  9.564268e-09\n",
       "5    2.421297e-10  2.663427e-09\n",
       "6    3.595062e-18  6.179013e-17\n",
       "7    8.634882e-04  5.052324e-03\n",
       "8    6.425397e-08  5.882944e-07\n",
       "9    3.623591e-05  2.557763e-04\n",
       "10   5.909617e-04  3.546691e-03\n",
       "11   3.732396e-05  2.631818e-04\n",
       "12   8.314858e-34  2.462477e-32\n",
       "13   2.734040e-11  3.209162e-10\n",
       "14   4.870036e-14  6.744474e-13\n",
       "15   1.293830e-31  3.596567e-30\n",
       "16   3.694966e-03  1.942064e-02\n",
       "17   6.605345e-51  3.413501e-49\n",
       "18   4.476906e-14  6.222414e-13\n",
       "19   2.535156e-05  1.820961e-04\n",
       "20   1.798356e-03  9.912200e-03\n",
       "21   5.482104e-04  3.302989e-03\n",
       "22   1.936689e-21  3.784900e-20\n",
       "23   2.638782e-04  1.657310e-03\n",
       "24   2.135358e-12  2.722228e-11\n",
       "25   2.324456e-30  6.150622e-29\n",
       "26   3.210560e-34  9.732799e-33\n",
       "27  3.321651e-191 1.967439e-188\n",
       "28   2.760745e-15  4.088026e-14\n",
       "29   7.034193e-72  6.298056e-70\n",
       "30   6.479635e-08  5.918528e-07\n",
       "⋮   ⋮             ⋮            \n",
       "119  1.133754e-13  1.542387e-12\n",
       "120  1.715073e-05  1.260120e-04\n",
       "121  1.338906e-38  4.863006e-37\n",
       "122  7.522957e-40  2.825696e-38\n",
       "123  1.045595e-23  2.205776e-22\n",
       "124  2.940189e-11  3.440647e-10\n",
       "125  1.512623e-17  2.488718e-16\n",
       "126  0.000000e+00  0.000000e+00\n",
       "127  2.278164e-27  5.447784e-26\n",
       "128  1.632778e-29  4.190796e-28\n",
       "129  8.969845e-51  4.574027e-49\n",
       "130  1.672711e-10  1.861253e-09\n",
       "131  7.373991e-34  2.192268e-32\n",
       "132  4.920331e-30  1.279951e-28\n",
       "133  2.480134e-14  3.478512e-13\n",
       "134  4.654712e-03  2.405455e-02\n",
       "135  7.075552e-05  4.825664e-04\n",
       "136  6.034788e-06  4.671111e-05\n",
       "137  1.162904e-61  8.140331e-60\n",
       "138  3.531118e-18  6.082687e-17\n",
       "139  3.506171e-19  6.307830e-18\n",
       "140  5.897128e-03  2.987361e-02\n",
       "141  4.424444e-09  4.430197e-08\n",
       "142 4.274129e-162 2.194053e-159\n",
       "143  3.431580e-36  1.129195e-34\n",
       "144  1.092093e-23  2.297573e-22\n",
       "145  5.629898e-43  2.269645e-41\n",
       "146  1.375070e-23  2.869387e-22\n",
       "147  4.279343e-42  1.655826e-40\n",
       "148  3.859810e-67  3.084533e-65"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "modules@meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "efd2faef-dcea-4a26-b037-0d7010ae3b0d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.1.1 (2021-08-10)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: CentOS Linux 8\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /home/rs619065/miniconda3/envs/r-4.1/lib/libopenblasp-r0.3.18.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
       "[8] base     \n",
       "\n",
       "other attached packages:\n",
       " [1] TFBSTools_1.32.0                  JASPAR2020_0.99.10               \n",
       " [3] Pando_0.4.2                       EnsDb.Hsapiens.v86_2.99.0        \n",
       " [5] ensembldb_2.18.2                  AnnotationFilter_1.18.0          \n",
       " [7] GenomicFeatures_1.46.1            AnnotationDbi_1.56.2             \n",
       " [9] BSgenome.Hsapiens.UCSC.hg38_1.4.4 BSgenome_1.62.0                  \n",
       "[11] rtracklayer_1.54.0                Biostrings_2.62.0                \n",
       "[13] XVector_0.34.0                    ggpubr_0.4.0                     \n",
       "[15] Nebulosa_1.4.0                    patchwork_1.1.1                  \n",
       "[17] cowplot_1.1.1                     dplyr_1.0.7                      \n",
       "[19] Signac_1.4.0                      SeuratObject_4.0.4               \n",
       "[21] Seurat_4.0.5                      ArchR_1.0.1                      \n",
       "[23] magrittr_2.0.1                    rhdf5_2.38.0                     \n",
       "[25] Matrix_1.3-4                      data.table_1.14.2                \n",
       "[27] SummarizedExperiment_1.24.0       Biobase_2.54.0                   \n",
       "[29] GenomicRanges_1.46.1              GenomeInfoDb_1.30.0              \n",
       "[31] IRanges_2.28.0                    S4Vectors_0.32.3                 \n",
       "[33] BiocGenerics_0.40.0               MatrixGenerics_1.6.0             \n",
       "[35] matrixStats_0.61.0                ggplot2_3.3.5                    \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] rappdirs_0.3.3              SnowballC_0.7.0            \n",
       "  [3] pbdZMQ_0.3-6                scattermore_0.7            \n",
       "  [5] R.methodsS3_1.8.1           tidyr_1.1.4                \n",
       "  [7] bit64_4.0.5                 knitr_1.36                 \n",
       "  [9] R.utils_2.11.0              irlba_2.3.3                \n",
       " [11] DelayedArray_0.20.0         rpart_4.1-15               \n",
       " [13] KEGGREST_1.34.0             RCurl_1.98-1.5             \n",
       " [15] generics_0.1.1              RSQLite_2.2.8              \n",
       " [17] RANN_2.6.1                  future_1.23.0              \n",
       " [19] tzdb_0.2.0                  bit_4.0.4                  \n",
       " [21] spatstat.data_2.1-0         xml2_1.3.2                 \n",
       " [23] httpuv_1.6.3                assertthat_0.2.1           \n",
       " [25] DirichletMultinomial_1.36.0 xfun_0.28                  \n",
       " [27] hms_1.1.1                   evaluate_0.14              \n",
       " [29] promises_1.2.0.1            fansi_0.5.0                \n",
       " [31] restfulr_0.0.13             progress_1.2.2             \n",
       " [33] caTools_1.18.2              dbplyr_2.1.1               \n",
       " [35] igraph_1.2.9                DBI_1.1.1                  \n",
       " [37] htmlwidgets_1.5.4           sparsesvd_0.2              \n",
       " [39] spatstat.geom_2.3-0         purrr_0.3.4                \n",
       " [41] ellipsis_0.3.2              ks_1.13.2                  \n",
       " [43] backports_1.4.0             annotate_1.72.0            \n",
       " [45] biomaRt_2.50.1              deldir_1.0-6               \n",
       " [47] sparseMatrixStats_1.6.0     vctrs_0.3.8                \n",
       " [49] SingleCellExperiment_1.16.0 ROCR_1.0-11                \n",
       " [51] abind_1.4-5                 cachem_1.0.6               \n",
       " [53] withr_2.4.2                 grr_0.9.5                  \n",
       " [55] ggforce_0.3.3               checkmate_2.0.0            \n",
       " [57] sctransform_0.3.2           GenomicAlignments_1.30.0   \n",
       " [59] prettyunits_1.1.1           mclust_5.4.8               \n",
       " [61] goftest_1.2-3               cluster_2.1.2              \n",
       " [63] IRdisplay_1.0               lazyeval_0.2.2             \n",
       " [65] seqLogo_1.60.0              crayon_1.4.2               \n",
       " [67] pkgconfig_2.0.3             slam_0.1-49                \n",
       " [69] tweenr_1.0.2                nlme_3.1-153               \n",
       " [71] ProtGenerics_1.26.0         nnet_7.3-16                \n",
       " [73] rlang_0.4.12                globals_0.14.0             \n",
       " [75] lifecycle_1.0.1             miniUI_0.1.1.1             \n",
       " [77] filelock_1.0.2              BiocFileCache_2.2.0        \n",
       " [79] dichromat_2.0-0             polyclip_1.10-0            \n",
       " [81] lmtest_0.9-39               ggseqlogo_0.1              \n",
       " [83] IRkernel_1.2.0.9000         carData_3.0-4              \n",
       " [85] Matrix.utils_0.9.8          Rhdf5lib_1.16.0            \n",
       " [87] zoo_1.8-9                   base64enc_0.1-3            \n",
       " [89] ggridges_0.5.3              png_0.1-7                  \n",
       " [91] viridisLite_0.4.0           rjson_0.2.20               \n",
       " [93] bitops_1.0-7                R.oo_1.24.0                \n",
       " [95] KernSmooth_2.23-20          rhdf5filters_1.6.0         \n",
       " [97] blob_1.2.2                  stringr_1.4.0              \n",
       " [99] parallelly_1.29.0           readr_2.1.0                \n",
       "[101] jpeg_0.1-9                  rstatix_0.7.0              \n",
       "[103] ggsignif_0.6.3              CNEr_1.30.0                \n",
       "[105] scales_1.1.1                memoise_2.0.0              \n",
       "[107] plyr_1.8.6                  ica_1.0-2                  \n",
       "[109] zlibbioc_1.40.0             compiler_4.1.1             \n",
       "[111] BiocIO_1.4.0                RColorBrewer_1.1-2         \n",
       "[113] fitdistrplus_1.1-6          Rsamtools_2.10.0           \n",
       "[115] listenv_0.8.0               pbapply_1.5-0              \n",
       "[117] htmlTable_2.3.0             Formula_1.2-4              \n",
       "[119] MASS_7.3-54                 mgcv_1.8-38                \n",
       "[121] tidyselect_1.1.1            stringi_1.7.5              \n",
       "[123] yaml_2.2.1                  latticeExtra_0.6-29        \n",
       "[125] ggrepel_0.9.1               grid_4.1.1                 \n",
       "[127] VariantAnnotation_1.40.0    fastmatch_1.1-3            \n",
       "[129] tools_4.1.1                 future.apply_1.8.1         \n",
       "[131] parallel_4.1.1              rstudioapi_0.13            \n",
       "[133] uuid_1.0-3                  TFMPvalue_0.0.8            \n",
       "[135] foreign_0.8-81              lsa_0.73.2                 \n",
       "[137] gridExtra_2.3               farver_2.1.0               \n",
       "[139] Rtsne_0.15                  digest_0.6.28              \n",
       "[141] shiny_1.7.1                 pracma_2.3.3               \n",
       "[143] qlcMatrix_0.9.7             motifmatchr_1.16.0         \n",
       "[145] Rcpp_1.0.7                  car_3.0-12                 \n",
       "[147] broom_0.7.10                later_1.3.0                \n",
       "[149] RcppAnnoy_0.0.19            httr_1.4.2                 \n",
       "[151] biovizBase_1.42.0           colorspace_2.0-2           \n",
       "[153] XML_3.99-0.8                tensor_1.5                 \n",
       "[155] reticulate_1.22             splines_4.1.1              \n",
       "[157] uwot_0.1.10                 RcppRoll_0.3.0             \n",
       "[159] spatstat.utils_2.2-0        plotly_4.10.0              \n",
       "[161] xtable_1.8-4                poweRlaw_0.70.6            \n",
       "[163] jsonlite_1.7.2              R6_2.5.1                   \n",
       "[165] Hmisc_4.6-0                 pillar_1.6.4               \n",
       "[167] htmltools_0.5.2             mime_0.12                  \n",
       "[169] glue_1.5.0                  fastmap_1.1.0              \n",
       "[171] BiocParallel_1.28.1         codetools_0.2-18           \n",
       "[173] mvtnorm_1.1-3               utf8_1.2.2                 \n",
       "[175] lattice_0.20-45             spatstat.sparse_2.0-0      \n",
       "[177] tibble_3.1.6                curl_4.3.2                 \n",
       "[179] leiden_0.3.9                gtools_3.9.2               \n",
       "[181] GO.db_3.14.0                survival_3.2-13            \n",
       "[183] repr_1.1.3                  docopt_0.7.1               \n",
       "[185] munsell_0.5.0               GenomeInfoDbData_1.2.7     \n",
       "[187] reshape2_1.4.4              gtable_0.3.0               \n",
       "[189] spatstat.core_2.3-1        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "728af060-703c-4969-9dd9-c032cb12b432",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
